src/tools/wasm-ctor-eval.cpp

```diff
@@ -0,0 +1,445 @@
+        }
+        break;
+      }
+      case i64: {
+        switch (load->bytes) {
+          case 1: return load->signed_ ? Literal((int64_t)doLoad<int8_t>(addr)) : Literal((int64_t)doLoad<uint8_t>(addr));
+          case 2: return load->signed_ ? Literal((int64_t)doLoad<int16_t>(addr)) : Literal((int64_t)doLoad<uint16_t>(addr));
+          case 4: return load->signed_ ? Literal((int64_t)doLoad<int32_t>(addr)) : Literal((int64_t)doLoad<uint32_t>(addr));
+          case 8: return load->signed_ ? Literal((int64_t)doLoad<int64_t>(addr)) : Literal((int64_t)doLoad<uint64_t>(addr));
+          default: WASM_UNREACHABLE();
+        }
+        break;
+      }
+      case f32: return Literal(doLoad<float>(addr));
+      case f64: return Literal(doLoad<double>(addr));
+      default: WASM_UNREACHABLE();
+    }
+  }
+
+  void store(Store* store, Address addr, Literal value) override {
+    // store to the memory segments
+    switch (store->valueType) {
+      case i32: {
+        switch (store->bytes) {
+          case 1: doStore<int8_t>(addr, value.geti32()); break;
+          case 2: doStore<int16_t>(addr, value.geti32()); break;
+          case 4: doStore<int32_t>(addr, value.geti32()); break;
+          default: WASM_UNREACHABLE();
+        }
+        break;
+      }
+      case i64: {
+        switch (store->bytes) {
+          case 1: doStore<int8_t>(addr, (int8_t)value.geti64()); break;
+          case 2: doStore<int16_t>(addr, (int16_t)value.geti64()); break;
+          case 4: doStore<int32_t>(addr, (int32_t)value.geti64()); break;
+          case 8: doStore<int64_t>(addr, value.geti64()); break;
+          default: WASM_UNREACHABLE();
+        }
+        break;
+      }
+      // write floats carefully, ensuring all bits reach memory
+      case f32: doStore<int32_t>(addr, value.reinterpreti32()); break;
+      case f64: doStore<int64_t>(addr, value.reinterpreti64()); break;
+      default: WASM_UNREACHABLE();
+    }
+  }
+
+  void growMemory(Address /*oldSize*/, Address newSize) override {
+    throw FailToEvalException("grow memory");
+  }
+
+  void trap(const char* why) override {
+    throw FailToEvalException(std::string("trap: ") + why);
+  }
+
+private:
+  // TODO: handle unaligned too, see shell-interface
+
+  template <typename T>
+  T* getMemory(Address address) {
+    // if memory is on the stack, use the stack
+    if (address >= instance->STACK_START) {
+      Address relative = address - instance->STACK_START;
+      if (relative + sizeof(T) > instance->STACK_SIZE) {
+        throw FailToEvalException("stack usage too high");
+      }
+      // in range, all is good, use the stack
+      return (T*)(&instance->stack[relative]);
+    }
+
+    // otherwise, this must be in the singleton segment. resize as needed
+    if (wasm->memory.segments.size() == 0) {
+      std::vector<char> temp;
+      wasm->memory.segments.push_back(
+        Memory::Segment(
+          wasm->allocator.alloc<Const>()->set(Literal(int32_t(0))),
```

Comment:
We could replace that `Segment` constructor with one that takes advantage of move semantics, which is probably what we want there anyway considering we `swap` the vectors, so having to `std::move` any lvalues we call it with makes it more obvious at the call site that the vector was mutated. Plus it lets us use rvalues for them.